can/interfaces/canalystii.py

```diff
@@ -69,132 +58,488 @@ class VCI_CAN_OBJ(Structure):
-                logger.error("VCI_StartCAN Error")
+        elif bitrate is not None:
+            if bitrate in TIMING_DICT:
+                Timing0, Timing1 = TIMING_DICT[bitrate]
+            else:
                 self.shutdown()
-                return
+                raise CanalystIIBitrateNotSupported(errno=bitrate)
+
+        elif None in (Timing0, Timing1):
+            self.shutdown()
+            raise CanalystIIInvalidTimings()
+
+        init_config.Timing0 = Timing0
+        init_config.Timing1 = Timing1
+
+        self.init_config = init_config
+
+        status = CANalystII.VCI_InitCAN(
+            VCI_USBCAN2,
+            device,
+            channel,
+            byref(self.init_config)
+        )
+
+        if status != STATUS_OK:
+            self.shutdown()
+            raise CANalystIIStartupError(
+                "VCI_InitCAN Error, incorrect channel number?",
+                status
+            )
+
+        status = CANalystII.VCI_StartCAN(
+            VCI_USBCAN2,
+            device,
+            channel
+        )
+
+        if status != STATUS_OK:
+            logger.error(
+                "VCI_StartCAN Error ({0}), device:{1} "
+                "channel:{2}".format(status, device, channel)
+            )
+            logger.info('receiving failed to start, only able to transmit.')
 
-    def send(self, msg, timeout=None):
+    def board_info(self) -> Dict:
         """
+        Adapter Information
+
+        :return: dictionary containing the following keys
+
+            * `hardware_version`: `int`
+            * `firmware_version`: `int`
+            * `driver_version`: `int`
+            * `library_version`: `int`
+            * `irq`: `int`
+            * `num_can_interfaces`: `int`
+            * `serial_number`: `str`
+            * `hardware_type`: `str`
+
+        :rtype: dict
+        """
+        return {key: value for key, value in self._b_info.items()}
+
+    def send(self, msg: Message, timeout=None):
+        """
+        Send CAN Frame
 
         :param msg: message to send
+        :type msg: :class:`can.Message`
+
         :param timeout: timeout is not used here
-        :return:
+        :type timeout: None
+
+        :return: `True` if frame was sent else `False`
+        :rtype: bool
         """
-        extern_flag = 1 if msg.is_extended_id else 0
-        raw_message = VCI_CAN_OBJ(
-            msg.arbitration_id,
-            0,
-            0,
-            1,
-            msg.is_remote_frame,
-            extern_flag,
-            msg.dlc,
-            (c_ubyte * 8)(*msg.data),
-            (c_byte * 3)(*[0, 0, 0]),
+
+        can_objs = (VCI_CAN_OBJ * 1)()
+
+        can_objs[0].ID = msg.arbitration_id
+        can_objs[0].ExternFlag = int(msg.is_extended_id)
+        can_objs[0].SendType = 1
+        can_objs[0].RemoteFlag = int(msg.is_remote_frame)
+        can_objs[0].DataLen = msg.dlc
+        for j in range(msg.dlc):
+            can_objs[0].Data[j] = msg.data[j]
```

Comment:
This will not work for remote frames. Don't use dlc to determine length of data.

Reply:
Ok this is fixed in the last commit.